tests/unit_tests/verRctNonSemanticsSimple.cpp

```diff
@@ -0,0 +1,265 @@
+    make_ctkey("4ee857d0ce17f66eca9c81eb326e404ceb50c8198248f2f827c440ee7aa0c0d7", "a8a9d61d4abbfb123630ffd214c834cc45113eaa51dd2f904cc6ae0c3c5d70e3")
+}};
+} // anonymous namespace
+
+TEST(verRctNonSemanticsSimple, tx_preconditions_1)
+{
+    // If this unit test fails, something changed about transaction deserialization / expansion or
+    // something changed about RingCT signature verification.
+
+    const cryptonote::transaction& tx = expand_transaction_from_bin_file_and_pubkeys
+        (tx1_file_name, tx1_input_pubkeys);
+    const rct::rctSig& rs = tx.rct_signatures;
+
+    ASSERT_EQ(1, tx.vin.size());
+    ASSERT_EQ(2, tx.vout.size());
+    const rct::key expected_sig_msg = rct::hash2rct(cryptonote::get_transaction_prefix_hash(tx));
+    ASSERT_EQ(expected_sig_msg, rs.message);
+    ASSERT_EQ(1, rs.mixRing.size());
+    ASSERT_EQ(16, rs.mixRing[0].size());
+    ASSERT_EQ(0, rs.pseudoOuts.size());
+    ASSERT_EQ(0, rs.p.rangeSigs.size());
+    ASSERT_EQ(0, rs.p.bulletproofs.size());
+    ASSERT_EQ(1, rs.p.bulletproofs_plus.size());
+    ASSERT_EQ(0, rs.p.MGs.size());
+    ASSERT_EQ(1, rs.p.CLSAGs.size());
+    ASSERT_EQ(1, rs.p.pseudoOuts.size());
+
+    ASSERT_TRUE(rct::verRctSemanticsSimple(rs));
+    ASSERT_TRUE(rct::verRctNonSemanticsSimple(rs));
+    ASSERT_TRUE(rct::verRctSimple(rs));
+    ASSERT_TRUE(rct::verRctNonSemanticsSimpleCached(tx));
+    ASSERT_TRUE(rct::verRctNonSemanticsSimpleCached(tx));
+}
+
+TEST(verRctNonSemanticsSimple, cached_serialize_assumption_1)
+{
+    // Hello, future visitors. If this unit test fails, than an important, consensus-related
+    // assumption has been broken and rct::verRctNonSemanticsSimpleCached() in file
+    // src/ringct/rctSigs.{h/cpp} needs to be reassessed for validity. The assumption is below:
+
+    const cryptonote::transaction& tx = expand_transaction_from_bin_file_and_pubkeys
+        (tx1_file_name, tx1_input_pubkeys);
+    const rct::rctSig& original_sig = tx.rct_signatures;
+    const auto original_sig_hash = rct_sig_to_hash(original_sig);
+
+    // These are the subtests most likely to fail. Fields 'message' and 'mixRing' are not serialized
+    // when sent over the wire, since they can be reconstructed from transaction data. Many images
+    // in prunable sig data e.g. 'p.CLSAGs[x].I' are not serialized as well for the same reason. 
+    // How signatures are serialized for the blockchain can be found in the methods
+    // rct::rctSigBase::serialize_rctsig_base and rct::rctSigPrunable::serialize_rctsig_prunable.
+    CACHED_SERIALIZATION_ASSUMPTION_SUBTEST(message.bytes[31]++)
+    CACHED_SERIALIZATION_ASSUMPTION_SUBTEST(mixRing.push_back({}))
+    CACHED_SERIALIZATION_ASSUMPTION_SUBTEST(mixRing[0].push_back({}))
+    CACHED_SERIALIZATION_ASSUMPTION_SUBTEST(mixRing[0][8].dest[10]--)
+    CACHED_SERIALIZATION_ASSUMPTION_SUBTEST(mixRing[0][15].mask[3]--)
+    //CACHED_SERIALIZATION_ASSUMPTION_SUBTEST(p.CLSAGs[0].I[2] += 101)
```

Comment:
Uncomment this now that it's fixed?

Reply:
It's not "fixed" really since right now it's still just testing `do_serialize` uniqueness. I think I'll have both a `do_serialize` uniqueness test and a full implementation mimic uniqueness test

Reply:
Okay I separated tests for `do_serialize()` uniqueness from the whole sig + tx uniqueness.